### PR TITLE
Implement a theming system (dark theme / night mode)

### PR DESCRIPTION
**Pull request by Jaifroid**
This backports (but with much improved and generic coding) a dark theme as described in #265.

I have implemented this as a generic theming system, though only three themes are currently implemented: `light` (the current default); `dark` (a theme for the app shell only); and `dark_invert` (a theme for the app shell and the content in the iframe). The screenshots below shows the `dark_invert` theme.

The app shell themes do not touch the document loaded in the iframe. the content theme of course has to attach a stylesheet, using DOM methods, after the document has loaded. It works in both jQuey and Service Worker modes.

Inversion is carried out by adding a simple inversion filter to the iframe element (`articleContent`), but we have to attach a stylesheet to the iframe document to re-invert images and videos (generically) so that they do not look like photographic negatives. There are two non-generic rules that are specific to Wikimedia ZIMs (the vast majority). This is practically inevitable: for discussion see https://github.com/openzim/mwoffliner/issues/1026 . If that proposal is implemented in the ZIM, it may be possible to remove one of the non-generic rules for future ZIMs. If a JS API is added to the ZIMs in the future, it will be possible to remove both non-generic rules.

**Image rendering**

image

**Configuration**

image

**Screenshot showing equations and search**

image

**Stackexchange**

image

### CODE REVIEW

**Comment by Jaifroid** [comment]
Working in FFOS (screenshot), but not in IE11. IE10+ does not support CSS filters and never will. Ironically, IE9 did support an inversion filter, but it was buggy and dropped from IE10+. As this feature does not prevent core functionality in the app, I don't think it's crucial to support IE11 in this particular case.

image

**Comment by mossroy** [comment]
That looks cool.
I agree with you about Internet Explorer : it's an optional and non-critical feature (even if quite appreciated, as you pointed), so never mind if such an old deprecated browser does not support it. It would be the same for Firefox OS (or other old browsers) if it didn't work on it.

As you would expect, I'm against putting some ZIM-specific code in our Kiwix client, so I don't like the mwoffliner-specific CSS classes.
I read the discussion on https://github.com/openzim/mwoffliner/issues/1026, and agree with @kelson42.
I also understand that there is currently no other way to have a 100% working night mode on these ZIM files.

Maybe a compromise would be to have 2 different "content themes" instead of a single one : 

- a generic one, that would only do what can be generic (like re-inverting the colors of images and videos). This one would not work 100% on current mwoffliner ZIM files (for mathML and SVG drawings). But it would be supported in the long term, and could take advantage of future (generic) mwoffliner improvements, like the JS API
- a "temporary" mwoffliner-specific one, that would add the mwoffliner-specific classes. This one would become unsupported as soon as the generic one could work 100%. But we would have to keep it for a while, for compatibility with people that kept old ZIM files.

We would need to find the right wording in the Settings section (and in the code). So that, in the future, a user that kept this "temporary" theme in its preferences would understand he/she should try to switch to the generic one if there are visual issues (with future ZIM files).

**Comment by Jaifroid** [comment]
Thanks for that feedback, @mossroy . I completely agree with all your points, and it's very easy to add further custom themes - I'll have a think about the wording. It would actually work with popular Wikipedia-specific dark themes such as https://github.com/StylishThemes/Wikipedia-Dark  - [CSS source code](https://github.com/StylishThemes/Wikipedia-Dark/blob/master/wikipedia-dark.user.css) too, but I think for our purposes what you suggest is good compromise.

**Comment by Jaifroid** [comment]
I've implemented these changes, with a pure theme (`dark_invert`) and one with Wikimedia workarounds (`dark_mwInvert`). I've also added some text to explain that appears if the latter is chosen. Finally, I've added a link that appears if the theme has changed for the user to display the article with the new theme applied, without reloading the landing page.

image

**Comment by Jaifroid** [comment]
No rush to review -- the code needs thorough field-testing, though I've done my usual checks in FFOS, Edge and Chromium, with Wikimedia, Stackexchange and Gutenberg ZIMs.

File: www/css/app.css [hunk 1]
```diff
@@ -75,6 +75,7 @@
     padding-left: 12px;
     padding-right: 12px;
     background: lightblue;
+    z-index: 1;
```
Comment (mossroy) [inline www/css/app.css]: I don't see how this relates to this PR

File: www/css/app.css [hunk 2]
```diff
@@ -89,6 +90,32 @@
...
+
+iframe._invert, iframe._mwInvert {
+    filter: invert(1) hue-rotate(180deg);
+}
+
+/* End of content themes */
+
+.hidden {
+    display: none;
+    margin: 5px;
```
Comment (mossroy) [inline www/css/app.css]: Why is this margin necessary? The class name is very generic and does not indicate that it adds a margin

File: www/css/app.css (hunk 1 above)
Comment (Jaifroid) [reply www/css/app.css]: @mossroy It turns out that the Article List (search results) have the wrong z-index, which is why the last search result in a long list can sometimes get hidden under the bottom bar. When the screen is inverted, this has the effect that the search results are hidden (they are no longer visible at all, though they can still be selected with the keyboard). The cure to both issues is to correct the z-index of the Article List. So it does need to be in this PR, even though it is a wider issue that this PR has revealed.

File: www/css/app.css (hunk 2 above)
Comment (Jaifroid) [reply www/css/app.css]: @mossroy The margin helps to position these elements in relation to each other. Originally I intended this to be a generic class for any hidden element, but it evolved into positioning as well. I'll rename to something related specifically to these links.

File: www/js/lib/uiUtil.js [hunk 3]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+        // If there is no ContentTheme or we are applying a different ContentTheme, remove any previously applied ContentTheme
+        if (oldContentTheme && oldContentTheme !== contentTheme) {
+            iframe.classList.remove(oldContentTheme);
+            if (kiwixJSSheet) {
+                kiwixJSSheet.disabled = true;
+                kiwixJSSheet.parentNode.removeChild(kiwixJSSheet);
+            }
+        }
+        // Apply the requested ContentTheme (if not already attached)
+        if (contentTheme && (!kiwixJSSheet || !~kiwixJSSheet.href.search(contentTheme))) {
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: @mossroy OK, I can make it more specific. I was being a bit lazy not wanting to construct a regex...

File: www/js/lib/uiUtil.js [hunk 4]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+            iframe.classList.add(contentTheme);
+            // Use an absolute reference becuase Service Worker needs this
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: Thanks

File: www/js/lib/uiUtil.js [hunk 5]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+            var prefix = document.location.href.replace(/index\.html(?:$|[#?].*$)/, '');
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: @mossroy The issue with SW mode is the usual one concerning ZIMs with subdirectories. If we are in a subdirectory (or even a document that has '/' in its filename, now treated by mwoffliner as a subdirectory), relative links to a stylesheet in the filesystem added after a document has loaded won't work. Since the browser can give us an absolute URL for index.html (NB this is index.html in our filesystem), I thought it best to use that. I believe the URL for the app will always either be `…...……/www/`, or `...………./www/index.html[?#….]`. Unless we ourselves rename index.html to something else.
I suppose there is the risk another dev may rename index.html and then this will break. So I'll check out your alternatives. Thanks for those! (I did consider using `uiUtil.deriveZimUrlFromRelativeUrl()`, but it is coded to work only with ZIM URLs.)

File: www/js/lib/uiUtil.js [hunk 6]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+                var link = doc.createElement('link');
+                link.setAttribute('id', 'kiwixJSTheme');
+                link.setAttribute('rel', 'stylesheet');
+                link.setAttribute('type', 'text/css');
+                link.setAttribute('href', prefix + 'css/kiwixJS' + contentTheme + '.css');
+                doc.head.appendChild(link);
+            }
+        }
+        // If we are in Config and a real document has been loaded already, expose return link so user can see the result of the change
+        if (/active/.test(document.getElementById('liConfigureNav').classList) &&
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: @mossroy I can add boundaries `\b` either side of the regex, so it only matches a whole word "active". We control the adding of "active" in this app, so I thought it wasn't much of a risk, but best to be safe!

File: www/js/lib/uiUtil.js [hunk 7]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+            !/Placeholder\sfor\sinjecting\san\sarticle/.test(doc.title)) {
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: OK. I'm a regex guy!!

File: www/js/lib/uiUtil.js [hunk 8]
```diff
@@ -227,12 +228,92 @@ define([], function() {
...
+            showReturnLink();
+        }
+    }
+
+    // Displays the return link and handles click event. Called by applyAppTheme()
+    function showReturnLink() {
+        var viewArticle = document.getElementById('viewArticle');
+        viewArticle.style.display = 'block';
+        viewArticle.addEventListener('click', function(e) {
+            e.preventDefault();
```
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: @mossroy Once #579 is merged, I'll rebase this PR and can then use the new functions (I think) for adding the swipe.
This is a part of #127, but doesn't reload the article, simply doesn't wipe it from the Home div (there's no need to reload to apply the theme, because it is added using DOM methods).
I still think the return link is useful to the user, because it orients them after what could be a somewhat startling style change, especially for new users. The link is removed as soon as it is used. I need to ensure it is also removed if a different article is loaded (I'll add a line to hide it to the `applyAppTheme` function).

[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: Regarding implementation of #127, I think that really needs #523 to be complete first, otherwise I fear we'll be coding the feature twice, once for the current UI and again for the more compact UI.

[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: PS I'd be happy to remove this in the future if it turns out that #127 elegantly supersedes it.

File: www/js/lib/uiUtil.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: `document.location.origin + document.location.pathname` works perfectly.

File: www/js/lib/uiUtil.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jaifroid) [reply www/js/lib/uiUtil.js]: I now use the much better `classList.contains()`.

**Comment by mossroy** [comment]
I've tested on a Firefox OS device : an old platform issue seems to be back when you choose one of the two "Dark+invert" themes : the article content can not be scrolled any more (see #175).
Except from that, it seems to work fine on it.
As it's an optional feature, and this bug seems platform-specific (it does not happen on the Firefox OS Simulator), I think we can simply ignore this bug.

**Comment by Jaifroid** [comment]
Thanks @mossroy . I have noticed at a number of points when developing Kiwix JS Windows a similar issue to #175 on Windows Mobile, which seemed to do with with the OS sometimes sending touch only through to the top-level container -- in our case this would be `<section id="search-article" role="region">` which contains the entire UI **and** the iframe, not just the search-article UI as its ID implies. It would occasionally happen when switching from Config (or About) to the articleContent, that the article could not be scrolled by touch. At various points I fiddled with z-indexes, but don't have a clear-cut solution.

Of course, the issue you notice could be simpler: the app freezes because of the extra cpu used in calculating colour inversions on all the elements in this case.

Once I've merged this, I'll see if I can reproduce a set of conditions on another slow mobile, because it may be revealing something about the code or UI design. If I can reproduce something, I'll open a separate issue.
